### DISCUSSION TITLE
Building Admittance matrix which includes generator transient reactance.

### DISCUSSION THREAD

**Discussion by kasheex** [CURENT/andes#556] [Modeling]
Dear Dr Cui

Could you please help me out how to build the following admittance matrices. I need to construct $\bar{Y}_{GG}$, $\bar{Y}_{BG}$ which includes generator's transient reactances. How can I import `x1q` and `x1d` parameters and use in `Line` model? Do I need to build a separate model or is it possible to modify `Line` model as `build_y()` function added by you in the same `class`.

<img width="508" alt="Screenshot 2024-07-07 at 8 13 40 PM" src="https://github.com/CURENT/andes/assets/57295204/e1d7d369-a71a-4df4-9e21-485a7291cc5f">

**Comment by kasheex** [comment]
I tried to build separate model as follows but it didn't work:

```
"""
Classes for calculating all the admittance matrices including generator admittances.
"""
import numpy as np

from andes.core.param import ExtParam
from andes.core.service import (BackRef, ConstService, ExtService, IdxRepeat,
                                NumReduce, NumRepeat, RefFlatten,)
from andes.core.var import Algeb, ExtAlgeb, ExtState
from andes.shared import spmatrix
from andes.models.line.line import Line


class YBUS(Line):
    """
    Admittance calculation
    """
    
    def __init__(self, system, config):
        Line.__init__(self, system, config)
        self.group = 'Calculation'
        
        self.SynGen = BackRef(info='Back reference to SynGen idx')  # `self.SynGen` is an IndexBase
        self.SynGenIdx = RefFlatten(ref=self.SynGen)  # flattened IndexBase

        self.index_bases.append(self.SynGenIdx)
        """
        adding new ExtParam
        """
        self.xq1 = ExtParam(model='SynGen', src='xq1', indexer=self.SynGenIdx,
                            )
        self.xd1 = ExtParam(model='SynGen', src='xd1', indexer=self.SynGenIdx,
                            )

    @property
    def build_ygg(self):
        """
        Building internal impedance matrix of the generator
        """
        nb = self.system.SynGen.n
        
        y1 = self.u.v * (-2*1j) / (self.xd1.v + self.xq1.v)

        # build self and mutual admittances into Ygg
        self.Ygg = spmatrix(y1, self.SynGenIdx.a, self.SynGenIdx.a, (nb, nb), 'z')
        
        return self.Ygg
```

**Comment by cuihantao** [comment]
Dear Kasheex,

Sorry that I wasn't able to get back to you earlier. The power injection formulation in ANDES does not build the machine transient/subtransient reactances into the Ybus matrix. 

You might need to do it manually. `Line` has a method to get the Ybus matrix. Write some custom code to iterate over the machines and add the elements.

**Comment by jinningwang** [comment]
My suggestion regarding the model implementation part would be:
1. develop separate scripts to build the desired matrices and store them into CSVs.
2. develop a new model to use `NumParm` to read the matrices to parameters, see more details in https://docs.andes.app/en/latest/groupdoc/SynGen.html#plbvfu1
3. Regarding loading the file, the `PLBVFU1` read time series data, but your potential implementation might only need once.
